### PR TITLE
add support to deploy image from the internal registry

### PR DESCRIPTION
**Pull request by karthikjeeyar**
Added support for deploy image from the internal registry.

Peek 2019-10-24 07-07

screenshot-localhost-9000-2019 10 24-07-21-18

Story: https://jira.coreos.com/browse/ODC-1452

### CODE REVIEW

File: frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+            items={getImageStreamTags(values.imageStream.image)}
+            title="Tag"
+            disabled={values.imageStream.image === ''}
+            fullWidth
+            required
+          />
+          <div className="imagestream-separator">:</div>
+        </div>
+      </div>
+      {values.registry === 'internal' &&
```
Comment (debsmita1) [inline frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx]: user enum here

File: frontend/packages/dev-console/src/components/import/image-search/ImageSearchSection.tsx [hunk 2]
```diff
@@ -1,13 +1,55 @@
...
+    >
+      <RadioButtonField
+        name="registry"
+        label="Image Section"
+        options={[
+          { label: 'Image name from external registry', value: 'external' },
+          { label: 'Image name from internal registry', value: 'internal' },
+        ]}
+      />
+      {values.registry === 'external' ? (
```
Comment (debsmita1) [inline frontend/packages/dev-console/src/components/import/image-search/ImageSearchSection.tsx]: I feel we can have an enum instead of writing `'external'`
```
export enum imageRegistry {
  External = 'external',
  Internal = 'internal',
};

or

export const imageRegistry = {
  External: {
    value: 'external',
    label: 'Image name from external registry',
  },
  Internal: {
    value: 'internal',
    label: 'Image name from internal registry',
  },
};

File: frontend/packages/dev-console/src/components/import/image-search/ImageStream.scss [hunk 3]
```diff
@@ -0,0 +1,5 @@
+.imagestream-separator {
```
Comment (invincibleJai) [inline frontend/packages/dev-console/src/components/import/image-search/ImageStream.scss]: we can follow convention as we have i.e `odc-imagestream-separator `

**Comment by invincibleJai** [comment]
@karthikjeeyar one observation, as per UXD content on radio selection show be just below it, currently it's shown after both radio buttons

Screenshot 2019-10-24 at 5 38 45 PM

File: frontend/packages/dev-console/src/components/import/DeployImage.tsx [hunk 4]
```diff
@@ -39,6 +37,12 @@ const DeployImage: React.FC<Props> = ({ namespace, projects, activeApplication }
     },
     name: '',
     searchTerm: '',
+    registry: 'external',
+    imageStream: {
+      image: '',
+      tag: '',
+      namespace: '',
+    },
```
Comment (joshuawilson) [inline frontend/packages/dev-console/src/components/import/DeployImage.tsx]: Why did you need to add this instead of using the `isi` or "image stream image"?

File: frontend/packages/dev-console/src/components/import/deployImage-submit-utils.ts [hunk 5]
```diff
@@ -146,8 +148,8 @@ export const createDeploymentConfig = (
             containerNames: [name],
             from: {
               kind: 'ImageStreamTag',
-              name: `${name}:${tag}`,
-              namespace,
+              name: `${imgName || name}:${tag}`,
```
Comment (joshuawilson) [inline frontend/packages/dev-console/src/components/import/deployImage-submit-utils.ts]: what is the difference between imageStream and ISI?

File: frontend/packages/dev-console/src/components/import/DeployImage.tsx (hunk 4 above)
Comment (karthikjeeyar) [reply frontend/packages/dev-console/src/components/import/DeployImage.tsx]: I am using these three fields(dropdowns) in UI and using these values an API call is made, which will set `isi` and `image` fields.

**Comment by karthikjeeyar** [comment]
@invincibleJai  @sahil143  aligned with UXD

screenshot-localhost-9000-2019 10 31-13-05-07

File: frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+      });
+  }, [
+    setFieldValue,
+    imageName.value,
+    imageTag.value,
+    imageNamespace.value,
+    values.application.name,
+    setFieldError,
+  ]);
+  const getImageStreamByNamespace = (ns: string) => {
```
Comment (sahil143) [inline frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx]: This can be extracted out as a utility

File: frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+    const imageStreamsList = {};
+    const isBuilderImageNamespace = ns === builderImagesNamespace.Openshift;
+    const imgStreams = isBuilderImageNamespace
+      ? (normalizeBuilderImages(imageStreams) as NormalizedBuilderImages)
+      : (imageStreams as K8sResourceKind[]);
+    _.each(imgStreams, (img: BuilderImage | K8sResourceKind) => {
```
Comment (sahil143) [inline frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx]: please use `reduce` here and return it directly, once extracted as a utility

File: frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+      const { name, namespace } = isBuilderImageNamespace
+        ? (img as BuilderImage).obj.metadata
+        : (img as K8sResourceKind).metadata;
+      if (namespace === ns) {
+        imageStreamsList[name] = name;
+      }
+    });
+    return imageStreamsList;
+  };
+  const getImageStreamTags = (name: string) => {
```
Comment (sahil143) [inline frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx]: This can be extracted out as a utility

File: frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    const tags = {};
+    const imageStream = _.find(imageStreams, ['metadata.name', name]);
+    name &&
+      !_.isEmpty(imageStream) &&
+      _.each(imageStream.status.tags, ({ tag }) => {
```
Comment (sahil143) [inline frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx]: same as above, `reduce` can be used here

File: frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+                the command:
+                <p>
+                  <code>
+                    oc policy add-role-to-user system:image-puller system:serviceaccount:
+                    {values.project.name}:default -n {imageNamespace.value}
+                  </code>
+                </p>
+              </Alert>
+            </div>
+          </div>
```
Comment (sahil143) [inline frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx]: It would be better if we can make it a sperate component. will look clean. WDYT?

**Comment by sahil143** [comment]
/lgtm
created this bug for review comments: https://jira.coreos.com/browse/ODC-2129
@karthikjeeyar

File: frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+            items={getImageStreamTags(values.imageStream.image)}
+            title="Tag"
+            disabled={values.imageStream.image === ''}
+            fullWidth
+            required
+          />
+          <div className="odc-imagestream-separator">:</div>
+        </div>
+      </div>
+      {values.registry === registryType.Internal &&
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx]: This condition will never be tru since value of registry in case of internal is being set as `"Image name from internal registry"`.

File: frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+  const projectList = {};
+
+  imageStreams.reduce((acc, { metadata: { namespace } }) => {
+    if (!acc[namespace]) {
+      acc[namespace] = namespace;
+    }
+    return acc;
+  }, projectList);
+
+  const [imageNamespace] = useField('imageStream.namespace');
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx]: You do not need to use `useField` here if you're just going to use `value` property of it. `useField` is helpful if you're trying to create a custom field and you need to use things like `field.value`, `field.onChange` and `field.onBlur`.

File: frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+  const [imageTag] = useField('imageStream.tag');
+  const [imageName] = useField('imageStream.image');
+  const { values, setFieldValue, setFieldError, initialValues } = useFormikContext<FormikValues>();
+  const [selectedProject, setselectedProject] = React.useState(imageNamespace.value);
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx]: You do not need to save the same values in local state as well as formik state.

File: frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+    const tags = {};
+    const imageStream = _.find(imageStreams, ['metadata.name', name]);
+    name &&
+      !_.isEmpty(imageStream) &&
+      _.each(imageStream.status.tags, ({ tag }) => {
+        tags[tag] = tag;
+      });
+    return tags;
+  };
+  React.useEffect(() => {
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx]: Again i think this is not necessary here. You can just use `values.something` along with `useEffect` hook to reset your form values.

File: frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+          />
+        </div>
+        <div className="col-lg-4 col-md-4 col-sm-4 col-xs-12">
+          <DropdownField
+            name="imageStream.image"
+            label="ImageStreams"
+            items={getImageStreamByNamespace(values.imageStream.namespace)}
+            title="ImageStreams"
+            fullWidth
+            disabled={values.imageStream.namespace === ''}
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx]: Somehow because of PF4 styles or something, it doesn't look like disabled.

File: frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx (hunk 12 above)
Comment (rohitkrai03) [reply frontend/packages/dev-console/src/components/import/image-search/ImageStream.tsx]: Its better to use `values`

**Comment by rohitkrai03** [comment]
@karthikjeeyar We have this bug - https://jira.coreos.com/browse/ODC-2129 to handle all the review comments as your laptop crashed. Since, the changes are functional I am going to approve it.
